src/core/Core.ml

```diff
@@ -205,6 +226,12 @@ let diseq = (=/=)
 let delay g st = RStream.from_fun (fun () -> g () st)
 
 let conj f g st = RStream.bind (f st) g
+
+let structural var rr k st =
+  match Term.var var with
+  | None -> success st
+  | Some v -> success { st with State.prunes = Term.VarMap.add v (Obj.magic (rr,k)) (State.prunes st) }
```

Comment:
Shouldn't you check the constraint when you first add it? 
What if it is violated in the current substitution?

Reply:
fixed.